crates/red_knot_test/src/lib.rs

```diff
@@ -121,6 +140,41 @@ fn run_test(db: &mut db::Db, test: &parser::MarkdownTest) -> Result<(), Failures
         })
         .collect();
 
+    // Create a custom typeshed `VERSIONS` file if none was provided.
+    if let Some(ref typeshed_path) = custom_typeshed_path {
+        if !has_custom_versions_file {
+            let versions_file = typeshed_path.join("stdlib/VERSIONS");
+            let contents = typeshed_files
+                .iter()
+                .map(|path| {
+                    let module_path = path.as_str().trim_end_matches(".pyi").replace("/", ".");
+                    format!("{module_path}: 3.8-\n")
+                })
+                .collect::<String>();
+            dbg!(&contents);
+            db.write_file(&versions_file, contents).unwrap();
+        }
+    }
+
+    Program::get(db)
+        .set_python_version(db)
+        .to(test.configuration().python_version().unwrap_or_default());
+    Program::get(db)
+        .set_python_platform(db)
+        .to(test.configuration().python_platform().unwrap_or_default());
+
+    Program::get(db)
+        .update_search_paths(
+            db,
+            &SearchPathSettings {
+                src_roots: vec![src_path],
+                extra_paths: vec![],
+                typeshed: custom_typeshed_path,
+                site_packages: SitePackages::Known(vec![]),
+            },
+        )
+        .expect("Failed to update search paths in TestDb");
+
```

Comment:
Nit: Should we just call `Program::update_from_settings`?

It has the advantage that it avoids setting values (and, therefore, bumping the salsa revision) if the values are unchanged.

Reply:
... and looks better. Thanks.